src/backends/torch/optim/ranger.cc

```diff
@@ -0,0 +1,229 @@
+  }
+
+  bool operator==(const RangerOptions &lhs, const RangerOptions &rhs)
+  {
+    return (lhs.lr() == rhs.lr())
+           && (std::get<0>(lhs.betas()) == std::get<0>(rhs.betas()))
+           && (std::get<1>(lhs.betas()) == std::get<1>(rhs.betas()))
+           && (lhs.eps() == rhs.eps())
+           && (lhs.weight_decay() == rhs.weight_decay())
+           && (lhs.rectified() == rhs.rectified())
+           && (lhs.lookahead() == rhs.lookahead())
+           && (lhs.adabelief() == rhs.adabelief())
+           && (lhs.gradient_centralization() == rhs.gradient_centralization())
+           && (lhs.lsteps() == rhs.lsteps()) && (lhs.lalpha() == rhs.lalpha());
+  }
+
+  void RangerOptions::serialize(torch::serialize::OutputArchive &archive) const
+  {
+    _TORCH_OPTIM_SERIALIZE_TORCH_ARG(lr);
+    _TORCH_OPTIM_SERIALIZE_TORCH_ARG(betas);
+    _TORCH_OPTIM_SERIALIZE_TORCH_ARG(eps);
+    _TORCH_OPTIM_SERIALIZE_TORCH_ARG(weight_decay);
+    _TORCH_OPTIM_SERIALIZE_TORCH_ARG(decoupled_wd);
+    _TORCH_OPTIM_SERIALIZE_TORCH_ARG(rectified);
+    _TORCH_OPTIM_SERIALIZE_TORCH_ARG(lookahead);
+    _TORCH_OPTIM_SERIALIZE_TORCH_ARG(adabelief);
+    _TORCH_OPTIM_SERIALIZE_TORCH_ARG(gradient_centralization);
+    _TORCH_OPTIM_SERIALIZE_TORCH_ARG(lsteps);
+    _TORCH_OPTIM_SERIALIZE_TORCH_ARG(lalpha);
+  }
+
+  void RangerOptions::serialize(torch::serialize::InputArchive &archive)
+  {
+    _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(double, lr);
+    _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(betas_t, betas);
+    _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(double, eps);
+    _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(double, weight_decay);
+    _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(bool, decoupled_wd);
+    _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(bool, rectified);
+    _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(bool, lookahead);
+    _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(bool, adabelief);
+    _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(bool, gradient_centralization);
+    _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(int, lsteps);
+    _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(double, lalpha);
+  }
+
+  bool operator==(const RangerParamState &lhs, const RangerParamState &rhs)
+  {
+    return ((lhs.step() == rhs.step())
+            && torch::equal(lhs.exp_avg(), rhs.exp_avg())
+            && torch::equal(lhs.exp_avg_sq(), rhs.exp_avg_sq()));
+  }
+
+  void
+  RangerParamState::serialize(torch::serialize::OutputArchive &archive) const
+  {
+    _TORCH_OPTIM_SERIALIZE_TORCH_ARG(step);
+    _TORCH_OPTIM_SERIALIZE_TORCH_ARG(exp_avg);
+    _TORCH_OPTIM_SERIALIZE_TORCH_ARG(exp_avg_sq);
+    _TORCH_OPTIM_SERIALIZE_TORCH_ARG(slow_buffer);
+  }
+
+  void RangerParamState::serialize(torch::serialize::InputArchive &archive)
+  {
+    _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(int64_t, step);
+    _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(torch::Tensor, exp_avg);
+    _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(torch::Tensor, exp_avg_sq);
+    _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(torch::Tensor, slow_buffer);
+  }
+
+  torch::Tensor Ranger::step(LossClosure closure)
```

Comment:
If it's inspired from libtorch Adam implementation for instance, I'd put a link to the original code, just in case if things change one day.

Reply:
ok